### ISSUE TITLE
[Bug]: Reloading logic divergence with models

### ISSUE THREAD

**Issue by tchedaTT** [tenstorrent/vllm#449]
### Your current environment

-

### 🐛 Describe the bug

vLLM passes reset_batch and reset_inputs as hints to the models for when then should reload the data from host to device.
However, different models implement supplementary logic and make this decision independently: see https://github.com/tenstorrent/tt-metal/blob/bc15d3423c517384a3fe41ec5ca6db1b359c714f/models/demos/llama3_70b_galaxy/tt/generator.py#L1386

This leads to at least a couple problems:
1. Stale inputs - sometimes the provided host inputs are stale under async decode. This **should** be ok, because vllm tries to match the async decode logic to the reloading logic. But if the model decides to reload when vllm doesnt recommend it, it will get stale inputs and, at best, duplicate some tokens.
2. Correctness in other models - if one model ships a fix for a missed case when inputs should be reloaded but aren't, other models stay incorrect in that scenario.
3. Sampling separation - to support https://github.com/tenstorrent/vllm/issues/277, we need to call device sampling as a separate step from vllm, because grammar may not be completed when we schedule the decode itself. To do that, we need to separate the logic for loading and reloading sampling params from loading/reloading other inputs - currently, it's intertwined.
4. Async decode correctness - currently only a few models (or maybe just llama70b) supports the steady-state async decode. When the contract is set and documented, there should be negligible overhead to enable it for other models.
5. Bringing up new models - because the contract isn't documented, it's a common scenario to miss it - things like slot_remap etc.

The solution is to keep all the reloading logic in vllm and make it match the steady-state logic.
Review should contain a small proof of why it matches and guarantees correct inputs at each stage, no extra tokens emitted from cancelled steps, etc.
The contract for when a generator is supposed to reload should be set and well documented.
Things that should be considered:
- reloading just the page table. Don't reload all other inputs then - I think there is a scenario where page table is guaranteed to be refreshed because the scheduler does it, and the other inputs aren't. Reloading everything at page table change would be a prohibitive performance drop
- slot_remap and changing the batch layout
- host sampling - loading new tokens from host with no changes to the batch layout and sampling params
- incrementing the position in the device sampling case
- prefill/decode switch - what needs to be reloaded then?
- any other optimizations and correctness fixes currently implemented in other models, such as llama. They need to be carefully verified, because they may be false leads.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by tchedaTT** [comment]
Another point to cover: Make sure the reloading decision considers the same scope as the draining decision, potential bugs like https://github.com/tenstorrent/vllm/pull/444 including the remaining one in the non-lane path

**Comment by tchedaTT** [comment]
Make sure that any reloading decision happens before the draining/overlap decision.
The current ordering is approximately:
apply_ready_completed_decode_steps()

if must_drain_pending_async_steps(
    can_attempt_steady_decode_from_scheduler(scheduler_output)
):
    wait_for_all_pending_async_steps()

model_input = build_model_input(scheduler_output)
# build_model_input calls _update_states()
[Source](https://github.com/tenstorrent/vllm/blob/1765019009fe867f085d511e626fd82e4c863d03/plugins/vllm-tt-plugin/src/vllm_tt_plugin/model_runner.py#L2093-L2118)
Notice that the drain decision happens before _update_states().
Suppose step N is pending, and the new scheduler output removes B:
Host before completion of N:
    A token/position: stale
    B token/position: stale
    C token/position: stale
    D token/position: stale

Device:
    A/B/C/D already advanced by step N
The pre-check may decide that overlap is allowed. Then _update_states() discovers that B disappeared, moves D into B’s row, and sets:
reset_batch = true
slot_remap = [0,3,2,3]
Because reset_batch instructs the model to reload batch inputs, it may reload stale host tokens and positions for A, C and D, overwriting the correct device-resident state. That can cause repeated/doubled tokens.

**Comment by tchedaTT** [comment]
Infodump from earlier agent exploration:
Some additions from the earlier cross-model survey:

- As a factual clarification, vLLM currently carries `reset_batch`; `reset_inputs` is generally derived or accepted inside individual model generators rather than being part of the common `TTModelInput` contract.

- The contract should distinguish independent reload domains:

  1. Decode-forward inputs such as token and current position.
  2. Page tables.
  3. Sampling parameters.
  4. Per-request sampling state, including RNG counters and prompt/output history used for penalties.

  A change in one domain should not automatically reload all the others.

- Reloading the whole batch is correctness-safe as long as vLLM guarantees that all supplied host inputs are current, authoritative, and belong to the same scheduler step. Partial reloads are primarily a performance optimization, especially for page-table-only changes. Under async device sampling, this invariant is important because an unexpected model-side reload can consume a one-step-stale host token.

- Host/device sampling transitions should be covered explicitly because token ownership differs:

  - Host sampling produces the next token on the host, so it must be loaded for the next decode.
  - Device sampling can feed the sampled token into the next device execution without a host reload.
  - When switching modes, the selected trace/input buffers may require a one-time refresh because they may not have been updated while the other mode was active.

- Sampling parameters normally do not change during a request. However, the batch-level sampling configuration changes when requests enter, leave, move, or reuse slots. Sampling parameters therefore need to be refreshed/remapped when the request-to-slot mapping changes, independently of whether the forward inputs need reloading.

- Slot remapping must also apply to per-request sampling state, not just model inputs. This includes RNG seed/counter state and prompt/output history used for repetition, presence, and frequency penalties. A reused slot must reset this state even if the padded batch shape did not change.

- Model adapters should either fully implement the common contract or declare that the relevant capability is unsupported. Arguments such as `slot_remap`, `prompt_tokens`, or `output_tokens` should not be silently accepted and ignored.

Separate/deferred sampling itself can remain outside the scope of this issue. However, keeping sampling-parameter/state refresh independent from forward-input reload is worth establishing now so that sampling can later be invoked separately without another interface redesign. Trace recapture does not need to be part of this contract; the concern here is refreshing the data used by an existing trace.

**Comment by tchedaTT** [comment]
### Two slot-remap failure modes across host/device sampling

`slot_remap` must be treated as a layout transaction, not as a device-sampler hint. There are two distinct ways an implementation can be wrong:

1. **vLLM does not deliver the remap on a host-sampling decode.**

   A layout change can happen on a step that falls back to host sampling (for example because of structured output, allowed/bad tokens, logits processors, or logprobs). Reloading tokens/positions from the host does not repair arbitrary model-owned state indexed by decode slot. Concrete examples are Qwen3.6's GDN recurrent/conv buffers and Qwen-VL's persistent per-slot RoPE deltas. If slot 3 moves to slot 0 and the remap is withheld because the device sampler is idle, the next forward reads slot 0 with state belonging to the old request.

2. **vLLM delivers and consumes the remap, but the adapter applies it only inside the active device-sampling path.**

   Moving `slot_remap` outside vLLM's `perform_device_sampling` condition is not sufficient by itself. If the adapter accepts the keyword during host sampling but only calls `seed_manager.apply_slot_remap(...)` from `sample_decode_on_device(...)`, vLLM may commit/clear the remap even though the dormant device sampling module never consumed it. A later host→device transition may reset that state and mask the bug, but relying on reset as an implicit substitute violates the explicit contract and becomes incorrect as soon as any per-slot sampling state survives or reset is incomplete.

The version-1 invariant should therefore be:

- vLLM composes one mapping where `new_slot[i] <- old_slot[slot_remap[i]]` and delivers it on the next decode in either sampling mode;
- before acknowledging that decode, the adapter applies the mapping exactly once to **every** persistent subsystem keyed by vLLM slot: forward/recurrent state, request-specific RoPE state, RNG/seed counters, penalty/history state, and any future slot-bound state;
- authoritative state rebuilt directly for the new layout may supersede remapping that domain, but this must be explicit so the same mapping is not applied twice;
- only after successful application/submission may vLLM commit the mapping;
- version-0 adapters retain their legacy device-sampling-only behavior during rollout, without receiving the version-1 guarantee.

This separates three responsibilities that must agree: **delivery by vLLM, complete application by the model adapter, and commit/acknowledgment only after application**.
